### PR TITLE
Add Python agent run trace

### PR DESCRIPTION
**Pull request by xTRam1**
## Summary

Adds structured tracing for custom Python agents. The SDK emits events for every stdout/stderr line, sub-agent call, extension action, and side-effect; the frontend ([NaradaAI/frontend#1592](https://github.com/NaradaAI/frontend/pull/1592)) consumes these events and renders them on the observability dashboard.

## What changes

**`narada-core`** — new Pydantic models
- `PythonAgentRunTrace` (step_type: `pythonAgentRun`) wrapping a run's full execution
- `PythonTraceEvent` discriminated union: `stdout`, `stderr`, `subAgentCall`, `extensionAction`, `sideEffect`
- Added to `ApaStepTrace`; `parse_action_trace` round-trips the new shape

**`narada-pyodide`** — chokepoint-based instrumentation so adding a new SDK method needs no trace-layer changes
- `window.py:dispatch_request()` wrapped → every `window.agent(...)` emits a `subAgentCall` event (covers success / error / timeout)
- `window.py:_run_extension_action()` wrapped → every extension action emits an `extensionAction` event
- `utils.download_file` / `utils.render_html` → `sideEffect` events
- All events flow through `_narada_emit_trace_event(json_str)`, a Pyodide global the frontend registers

**Payload safety** — per-action summaries in `_trace.summarize_request` bound the persisted JSON size (hard caps on free-form strings; structured subsets for `write_google_sheet`, screenshot, `agentic_selector`, etc.).

## Version bumps (coupled)

- `narada-core`: 0.0.17 → **0.0.18**
- `narada-pyodide`: 0.0.43 → **0.0.45a2**
- `narada`: 0.1.42 → **0.1.43** (repin `narada-core==0.0.18` only, no code change)

Why bump `narada` too: if `narada-core@0.0.18` (new variant) shipped without a matching `narada` bump, external Python SDK users would silently break on `parse_action_trace` ValidationError the moment a response they polled contained a nested Python custom agent run.

## Example

This custom Python agent:

```python
from narada import Agent, LocalBrowserWindow
import narada

window = LocalBrowserWindow()
await window.go_to_url(url="https://example.com", new_tab=True)

narada.render_html("<h1>Trace ordering test</h1>")
print("[A] after render_html")

narada.download_file("ordering-test.txt", "x")
print("[B] after download_file")

raise ValueError("intentional final traceback")
```

emits this trace (rendered by the frontend PR):

```json
{
  "step_type": "pythonAgentRun",
  "status": "error",
  "duration_ms": 16028,
  "events": [
    { "kind": "extensionAction", "ts_start": 1777063298934, "ts_end": 1777063302993,
      "action_name": "go_to_url",
      "request_summary": { "url": "https://example.com", "new_tab": true },
      "status": "success" },
    { "kind": "sideEffect", "ts": 1777063302994,
      "effect_type": "render_html", "description": "Rendered HTML in a new tab" },
    { "kind": "stdout", "ts": 1777063302995, "text": "[A] after render_html" },
    { "kind": "sideEffect", "ts": 1777063302995,
      "effect_type": "download_file", "description": "Downloaded file: ordering-test.txt" },
    { "kind": "stdout", "ts": 1777063302996, "text": "[B] after download_file" },
    { "kind": "stderr", "ts": 1777063303004,
      "text": "Traceback (most recent call last):\n  File \"<exec>\", line 11, in <module>\nValueError: intentional final traceback\n" }
  ],
  "error_message": "Traceback ..."
}
```

## Tests

38 unit tests in `packages/narada-pyodide/tests/test_trace.py`:
- Truncation edge cases (None, exact length, over cap)
- `summarize_request` / `summarize_response` per action type
- Event emitters: shape, timeout status, error-message truncation, optional-field omission
- End-to-end: every emitter's output round-trips through `parse_action_trace` into a valid `PythonAgentRunTrace`

Run (the `narada` / `narada-pyodide` namespace collision requires uninstalling narada first; see `packages/narada-pyodide/tests/README.md`):

```bash
uv pip uninstall narada
uv run --package narada-pyodide pytest packages/narada-pyodide/tests/
```

## Publish steps (manual, post-merge)

```bash
uv build --package narada-core    && uv publish --token <pypi-token>
uv build --package narada-pyodide && uv publish --token <pypi-token>
uv build --package narada         && uv publish --token <pypi-token>

git tag narada-core@0.0.18 narada-pyodide@0.0.45a2 narada@0.1.43
git push origin narada-core@0.0.18 narada-pyodide@0.0.45a2 narada@0.1.43
```

Frontend PR [NaradaAI/frontend#1592](https://github.com/NaradaAI/frontend/pull/1592) pins `narada-pyodide==0.0.45a2` and micropips it at Pyodide runtime, so the new versions must be on PyPI before that PR ships.

### CODE REVIEW

File: packages/narada-pyodide/src/narada/_trace.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    PrintMessageRequest,
+    ReadGoogleSheetRequest,
+    ReadGoogleSheetResponse,
+    WriteGoogleSheetRequest,
+)
+from pydantic import BaseModel
+
+if TYPE_CHECKING:
+    # Injected by the JavaScript harness at worker startup (see
+    # `frontend/src/lib/apa/python/python.worker.ts`). narada-pyodide is
```
Comment (xTRam1) [inline packages/narada-pyodide/src/narada/_trace.py]: Remove this file path

File: packages/narada-pyodide/src/narada/_trace.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    # only ever imported under a Pyodide worker that has registered this
+    # builtin; there is no non-Pyodide code path.
+    def _narada_emit_trace_event(event_json: str) -> None: ...
+
+
+# Hard caps on payload sizes carried in trace events. Values are large enough
```
Comment (xTRam1) [inline packages/narada-pyodide/src/narada/_trace.py]: Let's not put these hard caps here, I want to persist everything as much as possible.

File: packages/narada-pyodide/src/narada/_trace.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+_MAX_MESSAGE_CHARS = 500
+_MAX_ERROR_CHARS = 1000
+_MAX_QUERY_CHARS = 200
+
+# When a sub-agent's response includes its own action trace (for example, the
+# operator's step-by-step actions), we forward that trace one level deep so
+# the dashboard can expand it. We do not forward deeper nesting — Python
+# agents that delegate into other Python agents would otherwise produce
+# exponentially-sized persisted traces.
+_MAX_NESTED_ACTION_TRACE_DEPTH = 1
```
Comment (xTRam1) [inline packages/narada-pyodide/src/narada/_trace.py]: Why do we need this? Remove.

File: packages/narada-pyodide/src/narada/_trace.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    propagated to user code. ``default=str`` catches stray non-serialisable
+    values (timestamps, Pydantic models, numpy scalars) by stringifying them.
+    """
+    try:
+        _narada_emit_trace_event(json.dumps(event, default=str))  # noqa: F821
+    except Exception:  # noqa: BLE001 — broad by design; see docstring
+        _logger.warning("trace event emission failed", exc_info=True)
+
+
+def summarize_request(request: ExtensionActionRequest) -> dict[str, Any]:
```
Comment (xTRam1) [inline packages/narada-pyodide/src/narada/_trace.py]: If we are removing the hard cap things, do we still need these? We can remove this?

File: packages/narada-pyodide/src/narada/_trace.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+                request.fallback_operator_query, _MAX_QUERY_CHARS
+            ),
+        }
+    # ExtensionActionRequest is a closed union today. If a new variant is
+    # added without updating this function, we degrade gracefully to an empty
+    # summary rather than crashing the user's agent mid-run.
+    return {}
+
+
+def summarize_response(
```
Comment (xTRam1) [inline packages/narada-pyodide/src/narada/_trace.py]: Since we are removing the hard caps, we probably don't need these right?

File: packages/narada-core/src/narada_core/actions/models.py [hunk 6]
```diff
@@ -214,6 +221,99 @@ class OutputTrace(BaseModel):
...
+class PythonAgentRunTrace(BaseModel):
+    step_type: Literal["pythonAgentRun"] = "pythonAgentRun"
+    url: str
+    status: Literal["success", "error", "aborted"]
+    duration_ms: NonNegativeInt
+    events: list[PythonTraceEvent]
+    error_message: str | None = None
+    # Set by the runtime when it caps the number of buffered events (see
+    # `python.worker.ts`). Informational only; the dashboard surfaces it so
+    # users know their trace is partial.
```
Comment (xTRam1) [inline packages/narada-core/src/narada_core/actions/models.py]: We probably don't need this field, let's remove.

File: packages/narada-core/src/narada_core/actions/models.py [hunk 7]
```diff
@@ -214,6 +221,95 @@ class OutputTrace(BaseModel):
     description: str
 
 
+# ---------------------------------------------------------------------------
```
Comment (xTRam1) [inline packages/narada-core/src/narada_core/actions/models.py]: Move these to a model.py file inside a `tracing` folder.

**Review by zizhengtai — APPROVED** [review]
